### ISSUE TITLE
Guidance on setting Envoy runtime keys

### ISSUE THREAD

**Issue by sunjayBhatia** [projectcontour/contour#3452]
> @stevesloka  @skriss  Envoy still enforces max_program_size for re2 regexes via run time keys (defaults to 100): See the discussion here : https://github.com/envoyproxy/envoy/pull/11187.     it doesn't matter if using xDS v3 api.
whereas Contour since v1.10 have completely removed setting max_program_size for regexes in xDS route configuration, essentially breaking all users of regex path since Contour v1.10. 

> Any guidance on how to set Envoy's runtime key in the Envoy bootstrap config ?

_Originally posted by @ravilr in https://github.com/projectcontour/contour/issues/2933#issuecomment-791154676_

This issue seems like a generic one that could be expanded to allow support or at least guidance in docs for using Envoy runtime keys with Contour bootstrap, or even Contour dynamically sending config w/ RTDS

**Comment by sunjayBhatia** [comment]
@ravilr To start you could not rely on `contour bootstrap` and use a modified version of what that outputs as your Envoy bootstrap configuration (it isn't required that you use the output of that tool as your Envoy bootstrap config, you can certainly roll your own)

But it seems useful to have something in Contour that helps with setting up Envoy runtime config, even if it is just a guide/docs

**Comment by youngnick** [comment]
I think there are two things for us to do here:
- Fix the max_program_size as soon as we can, somehow. Istio did it by using the deprecated fields temporarily. (istio/istio#26153)
- Figure out a more general thing to do about runtime config of Envoy. I think that RTDS is probably the way to go here, so that you can flip things around without needing to restart Envoy.

**Comment by ravilr** [comment]
@sunjayBhatia  sure.  for others who stumble on to this due to the broken regex path in Ingress support, till an official Contour release gets this, you can either add the below layered runtime config snippet to envoy_config.json( passed as bootstrap config to envoy -c ):  

```
 "layered_runtime": {
     "layers": [
      {
       "name": "base",
       "static_layer": {
        "re2.max_program_size.error_level": 1048576
       }
      },
      {
       "name": "admin",
       "admin_layer": {}
      }
     ]
    }
```

or, if willing to run a patched version, below patch against contour-v1.10.1:
```
--- a/internal/envoy/v3/bootstrap.go
+++ b/internal/envoy/v3/bootstrap.go
@@ -33,10 +33,13 @@ import (
        envoy_matcher_v3 "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
        "github.com/golang/protobuf/proto"
        "github.com/golang/protobuf/ptypes/any"
+       _struct "github.com/golang/protobuf/ptypes/struct"
        "github.com/projectcontour/contour/internal/envoy"
        "github.com/projectcontour/contour/internal/protobuf"
 )
 
+const maxRegexProgramSize = 1 << 20
+
 // WriteBootstrap writes bootstrap configuration to files.
 func WriteBootstrap(c *envoy.BootstrapConfig) error {
        // Create Envoy bootstrap config and associated resource files.
@@ -149,6 +152,23 @@ func bootstrap(c *envoy.BootstrapConfig) ([]bootstrapf, error) {
 
 func bootstrapConfig(c *envoy.BootstrapConfig) *envoy_bootstrap_v3.Bootstrap {
        return &envoy_bootstrap_v3.Bootstrap{
+               LayeredRuntime: &envoy_bootstrap_v3.LayeredRuntime{
+                       Layers: []*envoy_bootstrap_v3.RuntimeLayer{{
+                               Name: "base",
+                               LayerSpecifier: &envoy_bootstrap_v3.RuntimeLayer_StaticLayer{
+                                       StaticLayer: &_struct.Struct{
+                                               Fields: map[string]*_struct.Value{
+                                                       "re2.max_program_size.error_level": {Kind: &_struct.Value_NumberValue{NumberValue: maxRegexProgramSize}},
+                                               },
+                                       },
+                               },
+                       }, {
+                               Name: "admin",
+                               LayerSpecifier: &envoy_bootstrap_v3.RuntimeLayer_AdminLayer_{
+                                       AdminLayer: &envoy_bootstrap_v3.RuntimeLayer_AdminLayer{},
+                               },
+                       }},
+               },
                DynamicResources: &envoy_bootstrap_v3.Bootstrap_DynamicResources{
                        LdsConfig: ConfigSource("contour"),
                        CdsConfig: ConfigSource("contour"),
```

And yes, i agree that Contour can expose a generic way of setting some of the runtime configurations(not all) in addition to `re2.*` settings. `overload.global_downstream_max_connections` is one such other setting:

`[warning][main] [source/server/server.cc:565] there is no configured limit to the number of allowed active connections. Set a limit via the runtime key overload.global_downstream_max_connections`

**Comment by sunjayBhatia** [comment]
For anyone following this issue, heres a PR to comment on that will be under review in the 1.22 release cycle: https://github.com/projectcontour/contour/pull/4387
